discovery_engine_core/core/src/engine.rs

```diff
@@ -72,85 +74,134 @@ where
             return Err(Error::NoStackOps);
         }
 
-        let mut stacks_data: HashMap<StackId, StackData> =
-            bincode::deserialize(state).map_err(Error::Deserialization)?;
+        let mut stacks_data = Self::deserialize(state)?;
 
         let stacks = stacks_ops
             .into_iter()
             .map(|ops| {
                 let id = ops.id();
                 let data = stacks_data.remove(&id).unwrap_or_default();
 
-                Stack::new(data, ops).map(|stack| (id, stack))
+                Stack::new(data, ops).map(|stack| (id, RwLock::new(stack)))
             })
             .collect::<Result<_, _>>()
             .map_err(Error::InvalidStack)?;
 
-        Ok(Engine { stacks, ranker })
+        Ok(Self { stacks, ranker })
     }
 
     /// Serializes the state of the `Engine`.
     ///
     /// The result can be used with [`Engine::new`] to restore it.
-    pub fn serialize(&self) -> Result<Vec<u8>, Error> {
-        let stacks_data: HashMap<&StackId, &StackData> = self
+    pub async fn serialize(&self) -> Result<Vec<u8>, Error> {
+        let state = self
             .stacks
             .iter()
-            .map(|(id, stack)| (id, &stack.data))
-            .collect();
+            .map(|(&id, stack)| async move {
+                bincode::serialize(&stack.read().await.data)
+                    .map(|data| (id, data))
+                    .map_err(Error::Serialization)
+            })
+            .collect::<FuturesUnordered<_>>()
+            .try_collect::<HashMap<_, _>>()
+            .await?;
+
+        bincode::serialize(&state).map_err(Error::Serialization)
+    }
 
-        bincode::serialize(&stacks_data).map_err(Error::Serialization)
+    /// Deserializes the state of the engine.
+    fn deserialize(state: &[u8]) -> Result<HashMap<StackId, StackData>, Error> {
+        bincode::deserialize::<HashMap<_, Vec<_>>>(state)
+            .map_err(Error::Deserialization)?
+            .into_iter()
+            .map(|(id, data)| {
+                bincode::deserialize(&data)
+                    .map(|data| (id, data))
+                    .map_err(Error::Deserialization)
+            })
+            .collect()
     }
 
     /// Returns at most `max_documents` [`Document`]s for the feed.
-    pub async fn get_feed_documents(&mut self, max_documents: u32) -> Result<Vec<Document>, Error>
-    where
-        R: Send + Sync,
-    {
-        Selection::new(BetaSampler)
-            .select(self.stacks.values_mut().collect(), max_documents)
+    pub async fn get_feed_documents(&self, max_documents: usize) -> Result<Vec<Document>, Error> {
+        Selection::new(BetaSampler, self.stacks.values())
+            .select(max_documents)
+            .await
             .map_err(|e| e.into())
     }
 
+    /// The ranker could rank the documents in a different order so we update the stacks with it.
+    async fn rank_stacks(&self) -> Result<(), Error> {
```

Comment:
Can't we leave the body of this in another function as it was? So to test it we don't have to create an engine.